### PR TITLE
[𝔸𝔻𝔻] 𝕍ℕ𝕠𝕕𝕖: 𝕋𝕒𝕓𝕝𝕖

### PR DESCRIPTION
**Pull request by Zinston**
Handle general table behavior.

Limitations:
- /!\ Does not include `unbreakable` so selection/backspace/delete/enter are momentarily broken in a table.
- Does not handle `<colgroup>`, `<tfoot>` and `<caption>`
- Add row/column only collapsed

### CODE REVIEW

File: packages/plugin-table/src/TableCellNode.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+            // Hand the managed cells over to the manager.
+            for (const managedCell of this.managedCells) {
+                managedCell.managerCell = newManager;
+                this._managedCells.delete(managedCell);
+            }
+        } else if (oldManager) {
+            // If we just removed this cell's manager, also remove this cell
+            // from the old manager's managed cells.
+            oldManager._managedCells.delete(this);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Now that I think about it, how would the system react if I set a manager on non-contiguous cells ?

File: packages/plugin-table/src/TableCellNode.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        }
+    }
+    /**
+     * Return the set of cells that this cell managed.
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Shouldn't it be "manage**s**" ?

File: packages/plugin-table/src/TableCellNode.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    /**
+     * Return the computed column span of this cell, in function of its managed
+     * cells.
+     */
+    get colspan(): number {
+        let colspan = 1;
+        for (const managedCell of this.managedCells) {
+            colspan = Math.max(colspan, managedCell.columnIndex - this.columnIndex + 1);
+        }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Wait, what ? Why isn't it `1 + this.managedCells.length` ?

File: packages/plugin-table/src/TableCellNode.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    /**
+     * Return the index of the row to which this cell belongs.
+     */
+    get rowIndex(): number {
+        return this.row.rowIndex;
+    }
+    /**
+     * Return the index of the column to which this cell belongs.
+     */
+    get columnIndex(): number {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: You sure do love getters, don't you ?

File: packages/plugin-table/src/TableDomParser.ts [hunk 5]
```diff
@@ -1,20 +1,129 @@
...
+import { TableCellNode } from './TableCellNode';
 
 export class TableDomParser extends AbstractParser<Node> {
     static id = 'dom';
     engine: DomParsingEngine;
 
-    predicate = (item: Node): boolean => {
-        return item instanceof Element && item.tagName === 'TABLE';
+    predicate = (item: Node): item is HTMLTableElement => {
+        return item.nodeName === 'TABLE';
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomParser.ts]: Why did you change this line ? The only problem it had was that `Element` should have been changed to `HTMLTableElement`.

File: packages/plugin-table/src/TableDomParser.ts [hunk 6]
```diff
@@ -1,20 +1,129 @@
...
         table.attributes = this.engine.parseAttributes(item);
-        const nodes = await this.engine.parse(...item.childNodes);
-        table.append(...nodes);
+
+        // Parse the contents of the table.
+        const children = await this.engine.parse(...item.childNodes);
+
+        // Build the grid.
+        const dimensions = this._getTableDimensions(item);
+        const parsedRows = children.filter(row => row.is(TableRowNode)) as TableRowNode[];
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomParser.ts]: What happens to the nodes that are not `TR` ? And compared to the DOM ?

File: packages/plugin-table/src/TableDomParser.ts [hunk 7]
```diff
@@ -1,20 +1,129 @@
...
+            const cells = grid[rowIndex];
+            let row = rows[rowIndex];
+            if (!row) {
+                row = new TableRowNode();
+            }
+            row.append(...cells);
+        }
+
+        // Append the rows to the table.
+        table.append(...rows);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomParser.ts]: It's a bit sad that we lose the grid from this point onward.

File: packages/plugin-table/src/TableDomParser.ts [hunk 8]
```diff
@@ -1,20 +1,129 @@
...
+    /**
+     * Return a tuple with the row length and the column length of the given DOM
+     * table element.
+     *
+     * @param domTable
+     */
+    _getTableDimensions(domTable: HTMLTableElement): [number, number] {
+        const domRows: HTMLTableRowElement[] = [];
+        for (const domChild of domTable.querySelectorAll('tr')) {
+            domRows.push(domChild);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomParser.ts]: Can't the result of `querySelectorAll` be transformed into an array rather than pushing it manually ?

File: packages/plugin-table/src/TableDomParser.ts [hunk 9]
```diff
@@ -1,20 +1,129 @@
...
+                while (grid[rowIndex][columnIndex]) {
+                    columnIndex += 1;
+                }
+
+                // Check traversing colspan and rowspan to insert placeholder
+                // cells where necessary. Consume these technical attributes.
+                const cellAttributes = cell.attributes['table-cell-attributes'];
+                delete cell.attributes['table-cell-attributes'];
+                const colspan = parseInt(cellAttributes.parsedColspan);
+                const rowspan = parseInt(cellAttributes.parsedRowspan);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomParser.ts]: Can't this function throw an error ?

File: packages/plugin-table/src/TableDomRenderer.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        const renderedChildren = await this.renderChildren(table);
+        for (const renderedChild of renderedChildren) {
+            for (const domChild of renderedChild) {
+                // `TableRowDomRenderer` returns a tbody or thead with a tr in
+                // it. Append the tr to `domHead` or `domBody` in function of
+                // what the children are. If we didn't receive a thead or a
+                // tbody, then the child was not parsed by `TableRowDomRenderer`
+                // (probably a technical node, like a script or style) and we
+                // need to simply append the child to `domTable`.
+                if (this._isTableSection(domChild)) {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomRenderer.ts]: This function is duplicated in the parser and renderer. I would have defined a `const tableSectionTagNames` or something like that in the Plugin itself and imported it here to use `includes` in the `if` rather than create a functon just for that.

File: packages/plugin-table/src/TableDomRenderer.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+                    const section = domChild.nodeName === 'THEAD' ? domHead : domBody;
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomRenderer.ts]: Especially if it's to re-test `THEAD` here...

File: packages/plugin-table/src/TableDomRenderer.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+                    // Transfer the attributes for the table section.
+                    for (const attribute of domChild.attributes) {
+                        section.setAttribute(attribute.name, attribute.value);
+                    }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableDomRenderer.ts]: Wait what ?

File: packages/plugin-table/src/TableNode.ts [hunk 13]
```diff
@@ -1,7 +1,81 @@
...
+     */
+    get name(): string {
+        return super.name + ': ' + this.rowCount + 'x' + this.columnCount;
+    }
+    /**
+     * Return an array of row nodes in this table.
+     */
+    get rowNodes(): TableRowNode[] {
+        return this.children(TableRowNode);
+    }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableNode.ts]: I would have named this one `rows`.

File: packages/plugin-table/src/TableNode.ts [hunk 14]
```diff
@@ -1,7 +1,81 @@
...
+    /**
+     * Return an array of rows in this table, as arrays of cells.
+     */
+    get rows(): TableCellNode[][] {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableNode.ts]: I would have named this one `grid` or `cells`.

File: packages/plugin-table/src/TableNode.ts [hunk 15]
```diff
@@ -1,7 +1,81 @@
...
+        const rows = [];
+        for (const row of this.rowNodes) {
+            const rowGrid = [];
+            for (let columnIndex = 0; columnIndex < this.columnCount; columnIndex += 1) {
+                rowGrid.push(this.rowNodes[row.rowIndex].cells[columnIndex]);
+            }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableNode.ts]: Why something like `rowGrid.push([...this.rowNodes[row.rowIndex].cells])` wouldn't work ?

File: packages/plugin-table/src/TableNode.ts [hunk 16]
```diff
@@ -1,7 +1,81 @@
...
+    }
+    /**
+     * Return an array of columns in this table, as arrays of cells.
+     */
+    get columns(): TableCellNode[][] {
+        const columns = [];
+        for (let columnIndex = 0; columnIndex < this.columnCount; columnIndex += 1) {
+            const columnGrid = [];
+            for (let rowIndex = 0; rowIndex < this.rowCount; rowIndex += 1) {
+                columnGrid.push(this.rowNodes[rowIndex].cells[columnIndex]);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableNode.ts]: Same here.

File: packages/plugin-table/src/TableRowDomParser.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * Parse a row node or a table section node.
+     *
+     * @param item
+     */
+    async parse(item: Element): Promise<TableRowNode[]> {
+        if (this._isTableSection(item)) {
+            return this.parseTableSection(item);
+        } else if (item.nodeName === 'TR') {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableRowDomParser.ts]: It's funny because you're re-playing the predicate but separately this time... ^^'

File: packages/plugin-table/src/TableRowDomRenderer.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        this.engine.renderAttributes(
+            row.attributes['table-section-attributes'] as Record<string, string>,
+            tableSection,
+        );
+        const renderedChildren = await this.renderChildren(row);
+        for (const renderedChild of renderedChildren) {
+            for (const domChild of renderedChild) {
+                domRow.appendChild(domChild);
+            }
+        }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableRowDomRenderer.ts]: I'm already tired of this `for/for` construction. Nothing related to your branch, though.

File: packages/plugin-table/src/Table.ts [hunk 19]
```diff
@@ -5,8 +5,136 @@ import { TableCellDomParser } from './TableCellDomParser';
...
+        },
+        {
+            pattern: 'CTRL+<Period>',
+            commandId: 'mergeCells',
+        },
+        {
+            pattern: 'CTRL+SHIFT+<Period>',
+            commandId: 'unmergeCells',
+        },
+    ];
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: Do we do something for these shortcuts ?

File: packages/plugin-table/src/Table.ts [hunk 20]
```diff
@@ -5,8 +5,136 @@ import { TableCellDomParser } from './TableCellDomParser';
...
+     *
+     * @param params
+     */
+    mergeCells(params: CommandParams): void {
+        const range = params.context.range;
+        const cells = range.targetedNodes(TableCellNode);
+        const managerCell = cells.shift();
+        for (const cell of cells) {
+            cell.managerCell = managerCell;
+        }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: I think a proper function would be more obvious, something like `setManagerCell` ? Or maybe override `mergeWith` to have a different behavior if the given node is a table cell ?

File: packages/plugin-table/src/TableNode.ts [hunk 21]
```diff
@@ -78,4 +78,130 @@ export class TableNode extends VElement {
...
+        referenceRow.before(newRow);
+        for (const cell of referenceRow.cells) {
+            const clone = cell.clone();
+            newRow.append(clone);
+
+            // Handle managers.
+            if (cell.managerCell) {
+                const manager = cell.managerCell;
+                if (manager.rowIndex === rowIndex + 1) {
+                    clone.managerCell = this.getCellAt(manager.rowIndex - 1, manager.columnIndex);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableNode.ts]: ?

File: packages/plugin-table/src/TableNode.ts [hunk 22]
```diff
@@ -78,4 +78,130 @@ export class TableNode extends VElement {
...
+                } else if (managerEndRow !== rowIndex) {
+                    // Take the manager cell of the above cell (extend rowspan),
+                    // only if said manager's rowspan is not ending with the
+                    // above cell.
+                    clone.managerCell = manager;
+                }
+            } else if (cell.rowspan > 1) {
+                // If the cell has a rowspan, extend it.
+                clone.managerCell = cell;
+            }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableNode.ts]: I think I don't get these `if/else if/else`...

Comment (Zinston) [reply packages/plugin-table/src/TableNode.ts]: OK 
image

**Comment by Zinston** [comment]
@dmo-odoo It's ready to be re-reviewed. I kept the most disruptive changes in separate commits appended to the PR to ease the review/adapt loop but of course the idea is to rearrange the commits later.

File: packages/plugin-table/src/TableCellDomRenderer.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+            for (const domChild of renderedChild) {
+                td.append(domChild);
+            }
+        }
+
+        // If a cell is empty, fill it to make it visible.
+        if (!cell.hasChildren()) {
+            const placeholderBr = document.createElement('BR');
+            td.append(placeholderBr);
+        }
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellDomRenderer.ts]: I would have done:
```typescript
if (cell.hasChildren()) {
    // the code at "Render the cell and its contents".
} else {
    // the br stuff
}
```

File: packages/plugin-table/src/TableCellNode.ts [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+            coordinatesRepr += ':(' + endRow + ', ' + endColumn + ')';
+        }
+        coordinatesRepr += '>';
+        return (
+            super.name +
+            coordinatesRepr +
+            (this.header ? ': header' : '') +
+            (this.isActive ? '' : ' (inactive)')
+        );
+    }
```
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: This looks very complicated.

File: packages/plugin-table/src/TableCellNode.ts [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+    set managerCell(newManager: TableCellNode | null) {
+        const oldManager = this._managerCell;
+        this._managerCell = newManager;
+
+        if (newManager) {
+            // If we just set a new manager for this cell, also add this cell to
+            // the new manager's managed cells.
+            newManager._managedCells.add(this);
+
+            // Copy the manager.
```
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: Well, you're not copying the manager as in "duplicating" him. You're copying the manager *attributes*.

File: packages/plugin-table/src/TableCellNode.ts [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+
+            // Hand the managed cells over to the manager.
+            for (const managedCell of this.managedCells) {
+                managedCell.managerCell = newManager;
+                this.__managedCells.delete(managedCell);
+            }
+        } else if (oldManager) {
+            // If we just removed this cell's manager, also remove this cell
+            // from the old manager's managed cells.
+            oldManager.__managedCells.delete(this);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Why is this in a `else if` ? I mean, you had an `oldManager`, I'm assignining you a `newManager`, don't you have to update `oldManager` anyway ?

File: packages/plugin-table/src/TableCellNode.ts [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+        }
+    }
+    /**
+     * Return the set of cells that this cell manages.
+     */
+    get managedCells(): Set<TableCellNode> {
+        return new Set(this.__managedCells);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Because of this getter and the hiding of the property, you are accessing the private property from the outside here: https://github.com/odoo-dev/jabberwock/pull/106/commits/c50eafebc0093984ccd685931b70485e0ddda044#diff-8696de8087e2f8c00c3b0a80ad497247R74

File: packages/plugin-table/src/TableCellNode.ts [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * Return true if this cell is active (ie not managed by another cell).
+     */
+    isActive(): boolean {
+        return !this.managerCell;
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Wait, if no cell is merged, none of them are `active` ?

File: packages/plugin-table/src/TableNode.ts (hunk 15 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/TableNode.ts]: I just realized you did not force push the branch, did you ?

File: packages/plugin-table/src/TableRowDomParser.ts [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+
+export class TableRowDomParser extends AbstractParser<Node> {
+    static id = 'dom';
+    engine: DomParsingEngine;
+
+    predicate = (item: Node): boolean => {
+        return (
+            item instanceof Element &&
+            (item.nodeName === 'THEAD' || item.nodeName === 'TBODY' || item.nodeName === 'TR')
+        );
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableRowDomParser.ts]: Now that we were able to check that together, the predicate return type could be `item is Element` and remove `item instanceof Element` in the function so that the double parenthesis can be removed.

File: packages/plugin-table/src/TableRowDomParser.ts (hunk 17 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/TableRowDomParser.ts]: What I meant is that it could be split into two different parsers, but honestly I'm not sure I would like that. Just something I noticed.

File: packages/plugin-table/src/Table.ts (hunk 19 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/Table.ts]: The toolbar *is* merged.

File: packages/plugin-table/src/Table.ts [hunk 30]
```diff
@@ -139,8 +144,51 @@ export class Table<T extends JWPluginConfig> extends JWPlugin<T> {
...
+     * Return true if the given array of cells forms a rectangle in the table
+     * grid.
+     *
+     * @param cells
+     */
+    _isRectangle(cells: TableCellNode[]): boolean {
+        // Add managed cells to the list.
+        for (const cell of cells.slice()) {
+            if (cell.managedCells.size) {
+                cells.push(...cell.managedCells);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: Wait, you're modifying the array you were given. Rather than do `.slice` in the `for`, you can do `cells = [...cells]`

File: packages/plugin-table/src/Table.ts [hunk 31]
```diff
@@ -139,8 +144,51 @@ export class Table<T extends JWPluginConfig> extends JWPlugin<T> {
...
+        // Compute the row/column index extrema.
+        const rowIndices = cells.map(cell => cell.rowIndex);
+        const columnIndices = cells.map(cell => cell.columnIndex);
+        const minRowIndex = Math.min(...rowIndices);
+        const minColumnIndex = Math.min(...columnIndices);
+        const maxRowIndex = Math.max(...rowIndices);
+        const maxColumnIndex = Math.max(...columnIndices);
+
+        // If a cell between the extrema cannot be found in the list, the
+        // selected cells do not form a quadrilateral.
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: quadrilateral

File: packages/plugin-table/src/TableCellNode.ts [hunk 32]
```diff
@@ -150,4 +120,58 @@ export class TableCellNode extends VNode {
...
+            managedCell.mergeWith(newManager);
+            this.__managedCells.delete(managedCell);
+        }
+
+        // Copy the manager's row if an entire row was merged
+        const row = this.ancestor(TableRowNode);
+        if (row) {
+            const areAllManagedByManager = row
+                .children(TableCellNode)
+                .every(cell => cell.managerCell === newManager);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Why not:
```typescript
const cells = row.children(TableCellNode);
const rowIsMerged = cells.every(cell => cell.managerCell === newManager);
```

File: packages/plugin-table/src/TableCellNode.ts [hunk 33]
```diff
@@ -150,4 +120,58 @@ export class TableCellNode extends VNode {
...
+            if (newManager.rowIndex !== this.rowIndex && areAllManagedByManager) {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: I don't understand the `rowIndex` check in this condition. How can `areAllManagedByManager` be true and the `rowIndex` condition not be true ?

File: packages/plugin-table/src/TableCellNode.ts [hunk 34]
```diff
@@ -150,4 +120,58 @@ export class TableCellNode extends VNode {
...
+                const managerRow = this.managerCell.ancestor(TableRowNode);
+                row.header = managerRow.header;
+                row.attributes = { ...managerRow.attributes };
+            }
+        }
+    }
+    unmerge(): void {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: doc

File: packages/plugin-table/src/TableCellNode.ts [hunk 35]
```diff
@@ -150,4 +120,58 @@ export class TableCellNode extends VNode {
...
+        if (this.__managerCell) {
+            // If we just removed this cell's manager, also remove this cell
+            // from the old manager's managed cells.
+            this.__managerCell.__managedCells.delete(this);
```
Comment (dmo-odoo) [inline packages/plugin-table/src/TableCellNode.ts]: Accessing private properties here as well.

File: packages/plugin-table/test/Table.test.ts [hunk 36]
```diff
@@ -3667,22 +3667,203 @@ describePlugin(Table, async testEditor => {
     describe('mergeCells', async () => {
         it('should merge two adjacent cells together', async () => {
             await testEditor(BasicEditor, {
-                contentBefore: '<table><tbody><tr><td>a[b</td><td>c]d</td></tr></tbody></table>',
+                /* eslint-disable prettier/prettier */
```
Comment (dmo-odoo) [inline packages/plugin-table/test/Table.test.ts]: boooooh

File: packages/plugin-table/src/Table.ts [hunk 37]
```diff
@@ -117,6 +140,132 @@ export class Table<T extends JWPluginConfig> extends JWPlugin<T> {
...
+     */
+    deleteRow(params: CommandParams): void {
+        const range = params.context.range;
+        const cell = range.start.ancestor(TableCellNode);
+        if (!cell) return;
+
+        const row = cell.ancestor(TableRowNode);
+        const nextRow = row.nextSibling(TableRowNode) || row.previousSibling(TableRowNode);
+        const nextCell = nextRow && nextRow.children(TableCellNode)[cell.columnIndex];
+        if (!nextCell) {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: Please switch these conditions to avoid negative conditions.

File: packages/plugin-table/src/Table.ts [hunk 38]
```diff
@@ -117,6 +140,132 @@ export class Table<T extends JWPluginConfig> extends JWPlugin<T> {
...
+     */
+    deleteColumn(params: CommandParams): void {
+        const range = params.context.range;
+        const cell = range.start.ancestor(TableCellNode);
+        if (!cell) return;
+
+        const column = cell.column;
+        const nextCell = cell.nextSibling(TableCellNode) || cell.previousSibling(TableCellNode);
+
+        if (!nextCell) {
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: Same here obviously.

File: packages/plugin-table/src/Table.ts [hunk 39]
```diff
@@ -117,6 +140,132 @@ export class Table<T extends JWPluginConfig> extends JWPlugin<T> {
...
+                } else if (!cell.isActive()) {
+                    // If this cell is inactive, unmerge it so its manager
+                    // doesn't believe it still manages it.
+                    cell.unmerge();
+                }
+                // Remove the cell.
+                cell.remove();
+
+                // The place where the range used to live was just demolished.
+                // Give it shelter within the next active cell.
```
Comment (dmo-odoo) [inline packages/plugin-table/src/Table.ts]: funny comment 👍 🍊

File: packages/plugin-table/src/TableCellNode.ts (hunk 35 above)
Comment (Zinston) [reply packages/plugin-table/src/TableCellNode.ts]: Same comment as above. Those private properties are built to be handled locally by merge/unmerge.

File: packages/plugin-table/test/Table.test.ts (hunk 36 above)
Comment (Zinston) [reply packages/plugin-table/test/Table.test.ts]: What? You prefer a gross and intrusive regular expression that removes the space? Or the whole html inline perhaps? :trollface:

File: packages/plugin-table/src/Table.ts (hunk 39 above)
Comment (Zinston) [reply packages/plugin-table/src/Table.ts]: :sunglasses:

File: karma.conf.js [hunk 40]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable */
```
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply karma.conf.js]: Manifestement, oui... ;-) cc @Goaman

File: packages/plugin-table/src/TableCellDomParser.ts [hunk 41]
```diff
@@ -0,0 +1,34 @@
+import { AbstractParser } from '../../core/src/AbstractParser';
+import { DomParsingEngine } from '../../plugin-dom/src/DomParsingEngine';
+import { TableCellNode } from './TableCellNode';
+
+export class TableCellDomParser extends AbstractParser<Node> {
+    static id = 'dom';
+    engine: DomParsingEngine;
+
+    predicate = (item: Node): item is HTMLTableCellElement => {
+        return item.nodeName === 'TD' || item.nodeName === 'TH';
```
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellDomParser.ts]: Yes we checked it together, this was an old comment.

File: packages/plugin-table/src/TableCellNode.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: Ok. However, the `this.isActive` bit cannot work, as `isActive` is a function and the property always exists. I suggest you re-test this function. :trollface:

File: packages/plugin-table/src/TableCellNode.ts (hunk 27 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: Well, if you do `<anything_other_than_"this">.__<private_property>` then yes, for me it is the outside. I'm wondering if it would be better or worse if instead of doing:
```typescript
managedCell.managerCell = manager;
```

We did:
```typescript
managerCell.manage(managedCell);
```

But I guess we're a bit screwed on that since we decided to use `mergeWith` as it would also need to write on the manager's private property from the outside...

File: packages/plugin-table/src/TableCellNode.ts (hunk 28 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: I guess I misread the code, my bad.

File: packages/plugin-table/src/TableCellNode.ts (hunk 33 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: But if we selected only part of a line, then `areAllManagedByManager` is going to be `false`, right ? So the entire condition will be `false` anyway without the need for the `rowIndex` condition, or did I miss something ?

File: packages/plugin-table/src/TableCellNode.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: No, it is not the same. When you copy/paste on your computer, the original is duplicated. So when you say you "copy the manager", I expect that we end up with a copy *of the manager* somewhere, which is not the case. You never create a *copy* of the manager. It's possible it's mostly my understanding which is biased by computing and "copy" might actually mean what you're saying it means (you know english better than I do), but I still think writing "Copy the manager's attributes and properties." instead of "Copy the manager." would not hurt anything and would improve the understandability for people like me whose english is highly influenced by computing, which is likely to be the case of most people who might stumble upon this comment.

File: packages/plugin-table/src/Table.ts (hunk 30 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/Table.ts]: Then I'd do `cells = [...cells]` above then `for (const cell of [...cells])` for the loop.

File: packages/plugin-table/test/Table.test.ts (hunk 36 above)
Comment (dmo-odoo) [reply packages/plugin-table/test/Table.test.ts]: Sincerely ? Yes I prefer the regex because at least the html in the test is easy to update. In your version, I have to manually manage the quotes and commas, while they do not even exist with the regex. With the regex, it's just like if you were in a proper html file, nothing else to worry about. You wrote a lot of tests though and it would be a waste of precious resources to have you update them (and to be fair, this regex is not even merged yet !) so we'll leave it at that for this PR.

File: packages/plugin-table/src/TableCellNode.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-table/src/TableCellNode.ts]: Fine, no problem.

File: packages/plugin-table/src/TableCellNode.ts (hunk 27 above)
Comment (Zinston) [reply packages/plugin-table/src/TableCellNode.ts]: Exactly...

You:
> if you do <anything_other_than_"this">.__<private_property> then yes, for me it is the outside

Code:
> `this.__managedCells`

Problem:
> ...

File: packages/plugin-table/src/Table.ts (hunk 30 above)
Comment (Zinston) [reply packages/plugin-table/src/Table.ts]: ... which is what I did :-)

File: packages/plugin-table/src/TableCellNode.ts (hunk 35 above)
Comment (Zinston) [reply packages/plugin-table/src/TableCellNode.ts]: Ok so here `__managedCells` is not accessed by `this`. but I believe my above comment still apply though. Whaddayathink?

File: packages/plugin-table/test/Table.test.ts (hunk 36 above)
Comment (Zinston) [reply packages/plugin-table/test/Table.test.ts]: Ok. Agree to disagree ^^

File: packages/plugin-table/src/TableCellNode.ts (hunk 27 above)
Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: Re-read my comment ! 🚎  The problem is not here, it is there: c50eafe#diff-8696de8087e2f8c00c3b0a80ad497247R74
I put the comment here because it is the existence of this getter that is somewhat the *cause* of the issue... ;-)

Comment (dmo-odoo) [reply packages/plugin-table/src/TableCellNode.ts]: I think we need to discuss these private properties by audio. I still need to work with christophe tomorrow but maybe we can squeeze like 10 minutes to discuss this.

**Comment by Zinston** [comment]
@dmo-odoo Done and reorganized.

**Comment by Zinston** [comment]
@dmo-odoo Ready.

**Comment by dmo-odoo** [comment]
@Zinston Je peux te demander d'ajouter un warning sur la page demo au dessus de la table pour expliquer que les suppressions ne sont pas supportées pour l'instant parce qu'on n'a pas encore fait unbreakable ?
